loki/analyse/util_polyhedron.py

```diff
@@ -0,0 +1,328 @@
+
+        NB: This can only deal with affine bounds, i.e. expressions that are
+            constant or can be reduced to a linear polynomial.
+
+        Upper bounds can be derived from this by multiplying left-hand side and
+        right-hand side with -1.
+
+        :param bound: the expression representing the lower bound.
+        :param list variables: the list of variable names.
+        :param int index: the index of the variable constrained by this bound.
+
+        :return: the pair ``(lhs, rhs)`` of the row in the matrix inequality.
+        :rtype: tuple(np.array, np.array)
+        """
+        supported_types = (sym.TypedSymbol, sym.MetaSymbol, sym.Sum, sym.Product)
+        if not (is_constant(bound) or isinstance(bound, supported_types)):
+            raise ValueError(f"Cannot derive inequality from bound {str(bound)}")
+        summands = accumulate_polynomial_terms(bound)
+        b = -summands.pop(1, 0)  # Constant term or 0
+        A = np.zeros([1, len(variables)], dtype=np.dtype(int))
+        A[0, index] = -1
+        for base, coef in summands.items():
+            if not len(base) == 1:
+                raise ValueError(f"Non-affine bound {str(bound)}")
+            A[0, variables.index(base[0].name.lower())] = coef
+        return A, b
+
+    @classmethod
+    def from_loop_ranges(cls, loop_variables, loop_ranges):
+        """
+        Create polyhedron from a list of loop ranges and associated variables.
+        """
+        assert len(loop_ranges) == len(loop_variables)
+
+        # Add any variables that are not loop variables to the vector of variables
+        variables = list(loop_variables)
+        variable_names = [v.name.lower() for v in variables]
+        for v in sorted(
+            FindVariables().visit(loop_ranges), key=lambda v: v.name.lower()
+        ):
+            if v.name.lower() not in variable_names:
+                variables += [v]
+                variable_names += [v.name.lower()]
+
+        n = 2 * len(loop_ranges)
+        d = len(variables)
+        A = np.zeros([n, d], dtype=np.dtype(int))
+        b = np.zeros([n], dtype=np.dtype(int))
+
+        for i, (loop_variable, loop_range) in enumerate(
+            zip(loop_variables, loop_ranges)
+        ):
+            assert loop_range.step is None or loop_range.step == "1"
+            j = variables.index(loop_variable.name.lower())
+
+            # Create inequality from lower bound
+            lhs, rhs = cls.generate_entries_for_lower_bound(
+                loop_range.start, variable_names, j
+            )
+            A[2 * i, :] = lhs
+            b[2 * i] = rhs
+
+            # Create inequality from upper bound
+            lhs, rhs = cls.generate_entries_for_lower_bound(
+                loop_range.stop, variable_names, j
+            )
+            A[2 * i + 1, :] = -lhs
+            b[2 * i + 1] = -rhs
+
+        return cls(A, b, variables)
+
+    @classmethod
+    def from_nested_loops(cls, nested_loops: list[Loop]):
```

Comment:
The type hint problem in Python 3.8 should go away by using
```python
from typing import List
```
and
```List[Loop]```

Reply:
eb8bb06d5ac8af02e35d02a8b9386c1bb7a8eb9b